### PR TITLE
feat(react): Recreate '/signin' page functionality in React

### PR DESCRIPTION
**Pull request by LZoog**
Because:
* We are porting Backbone flows to React

This commit:
* Adds signInRoutes and signUpRoutes config values for reference, adjust SignUp to conditionally redirect to React signin with router state, changes emailFromIndex to emailStatusChecked and passes status params along, accounts for various ways to receive email in Signin Container
* Modify local storage and currentAccount() data types and updates related files
* Makes Avatar component reusable by taking in the avatar data instead of useAccount, query avatar from profile on signin and display accordingly
* Creates query param model for signin, conditionally displays thirdPartyAuth to parity and UI tweaks to parity, "Use a different account" link fix with prefillEmail, reset PW link fix
* Adds functionality for cached credential signin and signin with password, handles error/success states

closes FXA-6488

---
You'll want to do most of your testing by going to [localhost:3030 with React params](http://localhost:3030/?forceExperiment=generalizedReactApp&forceExperimentGroup=react). **Note that some of the navigations take you to blank pages since they aren't set up in React yet**, what you should be testing is that it just takes you to the right path. Here's some flows worth testing, there's a lot of edge cases though so feel free to test other stuff too:
* Create an account, verify it, and then hit the above with the force experiment params. You should see the cached account view and not be prompted for the password and be able to sign in
* Go to `/clear`, then try to sign into the account. You should be prompted for a password and be able to sign in
* Test the above cases (non-cached and cached) with 2FA enabled
* Create an account, don't verify it, and try above cases (non-cached and cached)
* Hit [signin directly with showReactApp param](http://localhost:3030/signin?showReactApp=true) (forces the cache view). If nothing is in cache you should be redirected to `/`
* To see "session expired", sign in, then go to `/signin` and observe the cached view. In another tab, go to settings and change the password. Try to click "sign in" on the original tab.
* Hitting `/signin` with an `email` param that does not exist (should be redirected to `/signup` with that email)
* Navigate to "Use another account" and "reset password"

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/lib/router.js [hunk 1]
```diff
@@ -407,7 +407,19 @@ Router = Router.extend({
...
+        ...Url.searchParams(this.window.location.search),
+
+        email: this.user.get('emailFromIndex'),
+        hasLinkedAccount: this.user.get('hasLinkedAccount'),
+        hasPassword: this.user.get('hasPassword'),
+      });
+      // clear out because users can be navigated to `/signin` in flows that aren't
+      // directly from email-first/index, and we want an `email` query param to take
+      // precedence but want to read from local storage otherwise
+      // this.user.set('emailFromIndex', '');
```
Comment (LZoog) [inline packages/fxa-content-server/app/scripts/lib/router.js:421]: Thought I needed this, then discovered it was something else. I need to circle back to test some edge cases to see if we need it.

Probably hold on calling stuff out like this in review for now, I'll have it cleaned up Monday. 🙏

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    hasLinkedAccount:
+      queryParamModel.hasLinkedAccount || hasLinkedAccountFromLocationState,
+    hasPassword: queryParamModel.hasPassword || hasPasswordFromLocationState,
+  });
+  const { hasLinkedAccount, hasPassword } = accountStatus;
+
+  const nonCachedEmail = queryParamModel.email || emailFromLocationState;
+  let email = nonCachedEmail;
+  let sessionToken: hexstring | undefined;
+  let uid: hexstring | undefined;
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/container.tsx:92]: Will remove `uid` references in next push, I didn't end up needing it.

File: packages/fxa-content-server/server/lib/beta-settings.js [hunk 3]
```diff
@@ -77,6 +77,10 @@ const settingsConfig = {
   glean: { ...config.get('glean'), appDisplayVersion: config.get('version') },
   redirectAllowlist: config.get('redirect_check.allow_list'),
   sendFxAStatusOnSettings: config.get('featureFlags.sendFxAStatusOnSettings'),
+  showReactApp: {
+    signUpRoutes: config.get('showReactApp.signUpRoutes'),
+    signInRoutes: config.get('showReactApp.signInRoutes'),
```
Comment (LZoog) [inline packages/fxa-content-server/server/lib/beta-settings.js:82]: One quick note about this.

I added this to conditionally redirect to React or Backbone between SignIn/SignUp. In the Signup container, I used this accordingly. 

However, I didn't adjust the SignIn container to account for if SignUp routes were on or not because I didn't think it was worth the effort since SignUp should be rolled out when we roll SignIn out, but I left the config around (which I could remove). This just means **if we ever don't have React sign up rolled out but we _do_ have React sign in rolled out**, we're going to want a 1-pointer to also account for this in the signin container.

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    async (sessionToken: hexstring) => {
+      try {
+        // might need scope `profile:amr` for OAuth
+        const {
+          authenticationMethods,
+        }: { authenticationMethods: AuthenticationMethods[] } =
+          await authClient.accountProfile(sessionToken);
+
+        // after accountProfile data is retreived we must check verified status
+        // TODO: `uid` should never be undefined here, fix type
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/container.tsx:221]: ignore the `uid` comment, will remove comment with next push

File: packages/fxa-content-server/app/scripts/views/index.js [hunk 5]
```diff
@@ -266,6 +266,14 @@ class IndexView extends FormView {
       .then(({ exists, hasPassword, hasLinkedAccount }) => {
         const nextEndpoint = exists ? 'signin' : 'signup';
 
+        // Temporary hack for React work that allows us to pass the entered `email` as
+        // a param. When 'signup' and 'signin' flows are both finished and we're ready,
+        // we can convert the index page over and pass this along with router state
+        // instead of param.
+        this.user.set({ emailFromIndex: email });
```
Comment (dschom) [inline packages/fxa-content-server/app/scripts/views/index.js:273]: Just curious if there weould be a difference between `this.user.set({ emailFromIndex: email });` and `this.user.set('emailFromIndex', email);`?

**Review by vpomerleau — COMMENTED** [review]
A few early comments from poking around with unverified session forced globally.

I'm going to restart and look through again with normal settings.

File: packages/fxa-settings/src/pages/Signin/index.stories.tsx [hunk 6]
```diff
@@ -3,7 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import React from 'react';
-import Signin, { SigninProps } from '.';
+import Signin from '.';
 import AppLayout from '../../components/AppLayout';
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/index.stories.tsx:7]: Unused params after commenting out the sections below

File: packages/fxa-settings/src/pages/Signin/index.stories.tsx [hunk 7]
```diff
@@ -18,56 +18,54 @@ export default {
...
-}: SigninProps) => {
-  return (
-    <LocationProvider>
-      <AppLayout>
-        <Signin {...{ email, isPasswordNeeded, serviceName }} />
-      </AppLayout>
-    </LocationProvider>
-  );
-};
+// const SigninWithProvider = ({
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/index.stories.tsx:21]: Did you mean want to leave these commented out?

File: packages/fxa-settings/src/pages/Signin/index.test.tsx [hunk 8]
```diff
@@ -39,174 +39,174 @@ jest.mock('../../lib/glean', () => ({
...
-          // event was added so a success will follow a a submit.
-          // TODO
-          expect(GleanMetrics.cachedLogin.success).toHaveBeenCalledTimes(1);
-          expect(GleanMetrics.login.submit).toHaveBeenCalledTimes(0);
-        });
-      });
-    });
-  });
-});
+// describe('Signin', () => {
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/index.test.tsx:42]: Same question, is this supposed to be commented out?

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+      const options = {
+        verificationMethod: VerificationMethods.EMAIL_OTP,
+      };
+      // TODO in oauth ticket
+      //   // keys must be true to receive keyFetchToken for oAuth and syncDesktop
+      //   keys: isOAuth || isSyncDesktopV3,
+      //   service: service !== MozServices.Default ? service : undefined,
+      // };
+      try {
+        const { authPW } = await getCredentials(email, password);
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/container.tsx:165]: I'll have to add V2 key stretching support for this

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+  const cachedSigninHandler: CachedSigninHandler = useCallback(
+    async (sessionToken: hexstring) => {
+      try {
+        // might need scope `profile:amr` for OAuth
+        const {
+          authenticationMethods,
+        }: { authenticationMethods: AuthenticationMethods[] } =
+          await authClient.accountProfile(sessionToken);
+
+        // after accountProfile data is retreived we must check verified status
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/container.tsx:220]: typo, should be retrieved

File: packages/fxa-settings/src/pages/Signin/index.tsx [hunk 11]
```diff
@@ -2,117 +2,289 @@
...
+  const [signinLoading, setSigninLoading] = useState<boolean>(false);
+  const [bannerErrorText, setBannerErrorText] = useState<string>('');
+
+  const isOAuth = isOAuthIntegration(integration);
+  const isPocketClient = isOAuth && isClientPocket(integration.getService());
+  const isMonitorClient = isOAuth && isClientMonitor(integration.getService());
+  const hasLinkedAccountAndNoPassword = hasLinkedAccount && !hasPassword;
+
+  // We must use a ref because we may update this value in a callback
+  let isPasswordNeededRef = useRef(
```
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/index.tsx:73]: FWIW, this is also the case if I reset `forceGlobally` to false.

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+ * we can pass the param with router state. Since we already perform this account exists
+ * (account status) check on the Backbone index page, which is rate limited since it doesn't
+ * require a session token, we also temporarily pass email status params to 1) signal not to
+ * perform the check again but also because 2) these params are needed to conditionally
+ * display UI in signin. If no status params are passed and `email` is, or we read the
+ * email from local storage, we perform the check and redirect existing user emails to
+ * `/signup` to match content-server functionality.
+ */
+
+const SigninContainer = ({
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/container.tsx:59]: Are container tests going to be added in a follow up?

File: packages/fxa-settings/src/pages/Signin/interfaces.ts [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+  cachedSigninHandler: CachedSigninHandler;
+  sessionToken?: hexstring;
+  hasLinkedAccount: boolean;
+  hasPassword: boolean;
+  serviceName: MozServices;
+  avatarData: AvatarResponse | undefined;
+  avatarLoading: boolean;
+}
+
+export interface BeginSigninOptions {
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/interfaces.ts:43]: I'm not seeing any references to this.

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+          // the index page is converted to React, we need to run it in
+          // signup and signin for Sync
+          firefox.fxaCanLinkAccount({ email: queryParamModel.email });
+        }
+      }
+    })();
+  });
+
+  const { data: avatarData, loading: avatarLoading } =
+    useQuery<AvatarResponse>(AVATAR_QUERY);
```
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/container.tsx:148]: I believe this is how Backbone does it, I used to wonder why it sometimes didn't show the avatar until working on this. This just shows the default in this case which seems okay / to parity.

File: packages/fxa-settings/src/pages/Signin/index.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/index.tsx:73]: > FWIW, this is also the case if I reset forceGlobally to false.

Do we have docs for this?

I'm about to push an update to this ref that I noticed was incorrect while working on Storybook states, I might ask you to retest after. 👀

File: packages/fxa-settings/src/pages/Signin/index.test.tsx (hunk 8 above)
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/index.test.tsx:42]: No, I'm working on the tests and am about to push a stories update. I did this because the stack stopped building for me when I restarted and it got annoying having to fix the types for a build every time when props etc. were in-progress.

File: packages/fxa-settings/src/pages/Signin/container.tsx [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+    hasPassword: queryParamModel.hasPassword || hasPasswordFromLocationState,
+  });
+  const { hasLinkedAccount, hasPassword } = accountStatus;
+
+  const nonCachedEmail = queryParamModel.email || emailFromLocationState;
+  let email = nonCachedEmail;
+  let sessionToken: hexstring | undefined;
+  let uid: hexstring | undefined;
+  // only read from local storage if email isn't provided via query param or router state
+  if (!nonCachedEmail) {
```
Comment (vpomerleau) [inline packages/fxa-settings/src/pages/Signin/container.tsx:94]: I think we might not be hitting this and get redirected to index as soon as there is no "nonCachedEmail"

File: packages/fxa-settings/src/models/Account.ts [hunk 16]
```diff
@@ -79,9 +80,7 @@ export interface AttachedClient {
 export interface AccountData {
   uid: hexstring;
   displayName: string | null;
-  avatar: {
-    id: string | null;
-    url: string | null;
+  avatar: AccountAvatar & {
     isDefault: boolean;
```
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-settings/src/models/Account.ts:84]: It looks like we set this locally and don't pull it from GQL.

File: packages/fxa-settings/src/pages/Signin/container.tsx (hunk 12 above)
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/container.tsx:59]: Depends on how much I get done tomorrow, but the answer is probably 😅

File: packages/fxa-settings/src/pages/Signin/interfaces.ts (hunk 13 above)
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/interfaces.ts:43]: Good catch, was probably an artifact of an early on type I didn't end up using, will remove.

File: packages/fxa-settings/src/pages/Signin/index.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/index.tsx:73]: AFAIK, only the description in the config:
```
    forceGlobally: {
      doc: 'Force sign-in confirmation for all accounts',
      format: Boolean,
      default: false,
      env: 'SIGNIN_CONFIRMATION_FORCE_GLOBALLY',
    },
```

File: packages/fxa-settings/src/pages/Signin/container.tsx (hunk 15 above)
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/container.tsx:93]: What are your steps to reproduce?

File: packages/fxa-settings/src/pages/Signin/index.tsx [hunk 17]
```diff
@@ -2,117 +2,289 @@
...
+  );
 
-  // TODO in FXA-6488 use the integration's client id (instead of service name) to determine if client is Pocket or Monitor
-  const isPocketClient = serviceName === MozServices.Pocket;
-  const isMonitorClient = serviceName === MozServices.FirefoxMonitor;
-  const [error, setError] = useState('');
-  const [password, setPassword] = useState('');
-  const ftlMsgResolver = useFtlMsgResolver();
   const localizedPasswordFormLabel = ftlMsgResolver.getMsg(
     'password',
```
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/index.tsx]: Thanks for pointing all of these out!

File: packages/fxa-content-server/app/scripts/views/index.js (hunk 5 above)
Comment (LZoog) [reply packages/fxa-content-server/app/scripts/views/index.js:273]: I don't think there's a difference, so good catch on that being a little inconsistent. I'm guessing the former is usually for updating multiple values and the latter is for a single value, so I'll see if I can update all 3 with the first way.

File: packages/fxa-settings/src/pages/Signin/container.tsx (hunk 15 above)
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/container.tsx:93]: Nevermind me - I ended up in an odd state from creating an account, restarting my stack without signing out, and then going directly to `localhost:3030/signin` and `localhost:3030/signin?showReactApp=true`. In this situation, backbone suggests the previously signed in email (for an account that no longer exists), but react attempts to redirect to `/`, which redirects to backbone signin... Very much an edge case that we probably don't need to account for

File: packages/fxa-settings/src/pages/Signin/container.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/container.tsx:146]: Potentially, if https://github.com/mozilla/fxa/pull/16279 is merged in first, we could verify if the session is verified first, and if not skip the avatar query to avoid the must verify error? 🤔

[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/container.tsx:146]: Nice, I'll look at this in a follow up.

File: packages/fxa-settings/src/pages/Signin/index.test.tsx [hunk 18]
```diff
@@ -49,164 +124,222 @@ describe('Signin', () => {
...
+            expect(storeAccountData).toHaveBeenCalled();
+          });
 
-    // Pocket links should always open in a new window (announced by screen readers)
-    const pocketTermsLink = screen.getByRole('link', {
-      name: 'Terms of Service Opens in new window',
-    });
-    const pocketPrivacyLink = screen.getByRole('link', {
-      name: 'Privacy Notice Opens in new window',
+          // TODO in follow up, with container component tests
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/index.test.tsx:199]: Follow up here: https://mozilla-hub.atlassian.net/browse/FXA-9072

**Review by dschom — APPROVED** [review]
LGTM!

**Comment by vpomerleau** [comment]
Follow up: I filed https://mozilla-hub.atlassian.net/browse/FXA-9084 for the account chooser edge case I spotted in testing.

**Review by bcolsson — APPROVED** [review]
